library(R.utils)
library(numbat)
library(data.table)
library(Seurat)

argus <- (commandArgs(asValues=TRUE, excludeReserved=TRUE)[-1])
sample_name <- as.character(argus[1])
core_num <- as.numeric(argus[2])

set.seed(seed = 152727)


count_mat = Read10X_h5(paste0('../cell_ranger_output/',sample_name,'/outs/filtered_feature_bc_matrix.h5'))
df_allele = fread(paste0('gunzip -cq ../numbat/',sample_name,'/',sample_name,'_allele_counts.tsv.gz'))

# run
out = run_numbat(
    count_mat, # gene x cell integer UMI count matrix 
    ref_hca, # reference expression profile, a gene x cell type normalized expression level matrix
    df_allele, # allele dataframe generated by pileup_and_phase script
    genome = "hg38",
    t = 1e-5,
    ncores = core_num,
    plot = TRUE,
    out_dir = paste0('../numbat/',sample_name)
)
